### PR TITLE
rgw: Add coverity annotations for missing mutex locks

### PR DESCRIPTION
**Pull request by vedanshbhartia**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by yuvalif** [comment]
teuthology results: http://pulpito.ceph.com/yuvalif-2023-07-23_21:13:34-rgw-wip-yuval-test-gsco1-distro-default-smithi/

**Comment by yuvalif** [comment]
> teuthology results: http://pulpito.ceph.com/yuvalif-2023-07-23_21:13:34-rgw-wip-yuval-test-gsco1-distro-default-smithi/

all issues are kown:
* see: https://github.com/ceph/ceph/pull/52273#issuecomment-1647989729
* amqp crash: https://tracker.ceph.com/issues/61498

**Comment by cbodley** [comment]
where can i find information about these coverity reports? if these are real issues, we shouldn't suppress them

**Comment by yuvalif** [comment]
> where can i find information about these coverity reports? if these are real issues, we shouldn't suppress them

these are the issues in the coverity scan:
* https://scan5.scan.coverity.com/reports.htm#v58093/p10114/fileInstanceId=237731881&defectInstanceId=53132749&mergedDefectId=1405296
* https://scan5.scan.coverity.com/reports.htm#v58093/p10114/fileInstanceId=237731707&defectInstanceId=53017321&mergedDefectId=1510236
* https://scan5.scan.coverity.com/reports.htm#v58093/p10114/fileInstanceId=237731742&defectInstanceId=53158523&mergedDefectId=1510613
* https://scan5.scan.coverity.com/reports.htm#v58093/p10114/fileInstanceId=237731328&defectInstanceId=53076329&mergedDefectId=1511265
* https://scan5.scan.coverity.com/reports.htm#v58093/p10114/fileInstanceId=237731328&defectInstanceId=53076398&mergedDefectId=1511341

**Comment by cbodley** [comment]
> these are the issues in the coverity scan:

is there a way to access those without a login? if not, can we make a habit of copying the relevant info into tracker issues?

**Comment by yuvalif** [comment]
> > these are the issues in the coverity scan:
> 
> is there a way to access those without a login? if not, can we make a habit of copying the relevant info into tracker issues?

not possible without login (which requires approval).
ideally, we should publish the daily coverity scan results so that they are accesed publicly

**Comment by yuvalif** [comment]
jenkins test api

File: src/rgw/driver/rados/rgw_d3n_datacache.cc [hunk 1]
```diff
@@ -65,6 +65,7 @@ D3nDataCache::D3nDataCache()
 
 void D3nDataCache::init(CephContext *_cct) {
   cct = _cct;
+  // coverity[missing_lock:SUPPRESS]
```
Comment (yuvalif) [inline src/rgw/driver/rados/rgw_d3n_datacache.cc:68]: this wold be ok if `init()` is only called before any threads access `free_data_cache_size`
@mkogan1 can you please verify?

File: src/rgw/rgw_file_int.h [hunk 2]
```diff
@@ -226,6 +226,7 @@ namespace rgw {
       ~file();
     };
 
+    // coverity[missing_lock:SUPPRESS]
```
Comment (yuvalif) [inline src/rgw/rgw_file_int.h:229]: why is the lock supression before the definition of the struct?
which data member is not protected?

**Comment by yuvalif** [comment]
jenkins test api

File: src/rgw/driver/rados/rgw_d3n_datacache.cc (hunk 1 above)
Comment (yuvalif) [reply src/rgw/driver/rados/rgw_d3n_datacache.cc:68]: thanks!

Comment (yuvalif) [reply src/rgw/driver/rados/rgw_d3n_datacache.cc:68]: nvoked from `init_rados()` after ctor

**Review by yuvalif — APPROVED** [review]

**Comment by yuvalif** [comment]
jenkins test api

**Pull request merged**